### PR TITLE
add tests and missing assigner/equals

### PR DESCRIPTION
**Pull request by bd4**
Adding tests to prepare for SYCL commits, to make sure there are no regressions in CUDA and HIP.

### CODE REVIEW

**Comment by bd4** [comment]
All tests pass and examples work in CUDA. My HIP/ROCm software stack got broken and I am unable to get it to work even on master, so not sure if there are any regressions there... going to try to find another machine to test with.

There is a new nvcc warning introduced by the new launch tests. Pretty sure this is an existing problem exposed by the new tests, but not 100% sure.

**Comment by bd4** [comment]
Confirmed no regressions in HIP/ROCm support (and the new tests pass).

File: include/gtensor/gfunction.h [hunk 1]
```diff
@@ -128,54 +128,69 @@ inline void calc_shape(S& shape, const E& e)
...
+struct gtensor_inner_types<gfunction<F, E1, E2>>
+{
+  using space_type = space_t<expr_space_type<E1>, expr_space_type<E2>>;
+  constexpr static size_type dimension = helper::calc_dimension<E1, E2>();
+
+  using value_type =
+    decltype(std::declval<F>()(std::declval<expr_value_type<E1>>(),
+                               std::declval<expr_value_type<E2>>()));
+  using reference = value_type;
+  using const_reference = value_type;
```
Comment (germasch) [inline include/gtensor/gfunction.h:171]: I don't think this is anything new in this PR, but let me just note that this seems to be hackiness that should be addressed eventually.

Comment (bd4) [reply include/gtensor/gfunction.h:171]: Created issue for it: https://github.com/wdmapp/gtensor/issues/12

**Pull request merged**